server/admin/admin_server.go

```diff
@@ -72,12 +74,28 @@ func (s *Server) getTreeImpl(ctx context.Context, request *trillian.GetTreeReque
 func (s *Server) CreateTree(ctx context.Context, request *trillian.CreateTreeRequest) (*trillian.Tree, error) {
 	tree, err := s.createTreeImpl(ctx, request)
 	if err != nil {
-		return nil, errors.WrapError(err)
+		return nil, svrerrors.WrapError(err)
 	}
 	return tree, err
 }
 
 func (s *Server) createTreeImpl(ctx context.Context, request *trillian.CreateTreeRequest) (*trillian.Tree, error) {
+	// Test that the tree's private key exists. If it doesn't, generate it.
+	if _, err := s.registry.SignerFactory.NewSigner(ctx, request.GetTree()); err != nil {
+		if errors.ErrorCode(err) != errors.NotFound {
```

Comment:
I dislike the double-edged logic. I suggest the following:

* If a non-nil PrivateKey was supplied it must be valid or the RPC will fail
* If a nil PrivateKey was supplied we try to generate it (fail if we can't).

Generating a new key in case of errors may be surprising for callers. It'll also be hard to detect, given that we redact PKs from responses (it happens silently and without any feedback).